File: bindings/swift/Sources/RustPdf/Enums.swift
//
//  Enums.swift
//  RustPdf
//
//  Strongly-typed enums for the small-integer arguments the C ABI uses.
//

/// PDF/A conformance level (argument to ``Document/pdfa(_:)``).
public enum PdfaLevel: Int32, Sendable {
    case a1b = 0
    case a2b = 1
    case a2a = 2
    case a3b = 3
    case a3a = 4
    // PDF/A-4 (ISO 19005-4), based on PDF 2.0.
    case a4 = 5
    case a4e = 6
    case a4f = 7
}

/// Paragraph horizontal alignment (argument to ``Document/paragraph(font:size:x:y:width:text:align:)``).
public enum Align: Int32, Sendable {
    case left    = 0
    case right   = 1
    case center  = 2
    case justify = 3
}

/// Embedded-file relationship for PDF/A-3 attachments
/// (`/AFRelationship`; argument to ``Document/attachFile(name:mime:data:relationship:description:)``).
public enum AFRelationship: Int32, Sendable {
    case source      = 0
    case data        = 1
    case alternative = 2
    case supplement  = 3
    case unspecified = 4
}

/// Document encryption cipher (argument to ``EditableDoc/encrypt(method:user:owner:readOnly:)``).
public enum Encryption: Int32, Sendable {
    case rc4    = 0
    case aes128 = 1
    case aes256 = 2
}

/// ZUGFeRD / Factur-X invoice profile (argument to ``Document/facturx(_:profile:)``).
public enum FacturxProfile: Int32, Sendable {
    case minimum  = 0
    case basicWL  = 1
    case basic    = 2
    case en16931  = 3
    case extended = 4
}

/// DocMDP certification level applied by the first (certifying) signature
/// (``SigningOptions/certify``).
public enum Certify: Int32, Sendable {
    /// Not a certifying signature.
    case none = 0
    /// `/P 1` — no changes permitted after signing.
    case locked = 1
    /// `/P 2` — form-filling and signing permitted.
    case forms = 2
    /// `/P 3` — form-filling, signing and annotations permitted.
    case formsAndAnnotations = 3
}

/// What the `y` coordinate of a positioned text stamp means
/// (``EditableDoc/placeText(_:_:_:_:size:color:rotationDeg:align:fontId:anchor:)``
/// and the block anchor of ``EditableDoc/placeParagraph(_:_:_:_:_:size:color:align:fontId:maxHeight:lineHeight:anchor:rotationDeg:)``).
/// `baseline` is the historical default for `placeText`; `top` hangs the text
/// from `y` (baseline at `y − ascent × size`, legacy fixed-position layout
/// semantics); `bottom` rests the descender line on `y`. The `line*` cases use
/// the **layout line box** (OS/2 win metrics — or typo × 1.2 — plus the legacy engine's
/// default half-leading of 0.21 em) instead of the raw ascent/descent, matching
/// legacy layout engines line placement exactly. Ascent/descent come from the selected font
/// (embedded font metrics, or Helvetica AFM).
public enum VerticalAnchor: Int32, Sendable {
    case baseline   = 0
    case top        = 1
    case bottom     = 2
    /// Top of the layout line box.
    case lineTop    = 3
    /// Bottom of the layout line box.
    case lineBottom = 4
}

/// Vertical alignment of the text line inside a
/// ``EditableDoc/maskedText(_:_:_:_:_:_:size:textColor:bgColor:align:fontId:valign:padding:)``
/// box. `middle` (the historical default) centers the cap-height block; `top`
/// hangs the line from the top edge (baseline at `y + height − ascent × size`,
/// top line-alignment semantics of rectangle-based text APIs); `bottom` rests the descender
/// line on the bottom edge.
public enum VerticalAlign: Int32, Sendable {
    case top    = 0
    case middle = 1
    case bottom = 2
}

/// Coordinate space of the positioned stamping primitives (`fillRect`,
/// `placeText`, `maskedText`, `placeParagraph`, `drawImage`) — set via
/// ``EditableDoc/setStampSpace(_:)``. `visible` (historical default):
/// coordinates in the page's displayed space, compensating `/Rotate` so a
/// `rotationDeg = 0` stamp reads upright on screen. `media`: raw PDF user
/// space (legacy fixed-position layout semantics) — no
/// composition with the page's `/Rotate` or crop offset; `rotationDeg` is the
/// baseline angle in media space. Watermarks and redaction are unaffected.
public enum StampSpace: Int32, Sendable {
    case visible = 0
    case media   = 1
}

/// How a rotated image is anchored at `(x, y)`
/// (``EditableDoc/drawImage(_:image:x:y:width:height:rotationDeg:anchor:)``).
/// `corner` (default): the image's own lower-left corner — the image sweeps
/// around it when rotated. `boundingBox`: the rotated image's bounding box
/// lands with its lower-left at `(x, y)` (bounding-box layout semantics — pixels
/// always at/above/right of the anchor).
public enum ImageAnchor: Int32, Sendable {
    case corner      = 0
    case boundingBox = 1
}

/// The PDF version written in the header (argument to ``Document/setVersion(_:)``).
public enum PdfVersion: Int32, Sendable {
    case v14 = 0
    case v15 = 1
    case v17 = 2
    case v20 = 3
}

File: bindings/rust/src/enums.rs
//! Strongly-typed enums mirroring the small-int arguments the C ABI takes.

/// PDF/A conformance level. Passed to [`crate::Document::pdfa_level`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PdfaLevel {
    /// PDF/A-1b.
    A1b,
    /// PDF/A-2b.
    A2b,
    /// PDF/A-2a (also enables tagging).
    A2a,
    /// PDF/A-3b.
    A3b,
    /// PDF/A-3a (also enables tagging).
    A3a,
    /// PDF/A-4 (ISO 19005-4), based on PDF 2.0.
    A4,
    /// PDF/A-4e (engineering).
    A4e,
    /// PDF/A-4f (embedded files).
    A4f,
}

impl PdfaLevel {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::A1b => 0,
            Self::A2b => 1,
            Self::A2a => 2,
            Self::A3b => 3,
            Self::A3a => 4,
            Self::A4 => 5,
            Self::A4e => 6,
            Self::A4f => 7,
        }
    }
}

/// Paragraph alignment. Passed to [`crate::Document::paragraph`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Align {
    Left,
    Right,
    Center,
    Justify,
}

impl Align {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Left => 0,
            Self::Right => 1,
            Self::Center => 2,
            Self::Justify => 3,
        }
    }
}

/// Embedded-file relationship (PDF/A-3). Passed to [`crate::Document::attach_file`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum AFRelationship {
    Source,
    Data,
    Alternative,
    Supplement,
    Unspecified,
}

impl AFRelationship {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Source => 0,
            Self::Data => 1,
            Self::Alternative => 2,
            Self::Supplement => 3,
            Self::Unspecified => 4,
        }
    }
}

/// Encryption method. Passed to [`crate::EditableDoc::encrypt`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Encryption {
    /// RC4 128-bit.
    Rc4_128,
    /// AES-128.
    Aes128,
    /// AES-256 (R6).
    Aes256,
}

impl Encryption {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Rc4_128 => 0,
            Self::Aes128 => 1,
            Self::Aes256 => 2,
        }
    }
}

/// Factur-X / ZUGFeRD profile. Passed to [`crate::Document::facturx`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FacturxProfile {
    /// MINIMUM profile.
    Minimum,
    /// BASIC WL (without lines) profile.
    BasicWl,
    /// BASIC profile.
    Basic,
    /// EN 16931 (COMFORT) profile.
    En16931,
    /// EXTENDED profile.
    Extended,
}

impl FacturxProfile {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Minimum => 0,
            Self::BasicWl => 1,
            Self::Basic => 2,
            Self::En16931 => 3,
            Self::Extended => 4,
        }
    }
}

/// DocMDP certification level applied by a certifying signature (the first
/// signature only). Passed via [`crate::SigningOptions::certify`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum Certify {
    /// Not a certifying signature (an ordinary approval signature).
    #[default]
    None,
    /// `/P 1` — no changes permitted after signing.
    Locked,
    /// `/P 2` — form-filling and signing permitted.
    Forms,
    /// `/P 3` — form-filling, signing and annotations permitted.
    FormsAndAnnotations,
}

impl Certify {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::None => 0,
            Self::Locked => 1,
            Self::Forms => 2,
            Self::FormsAndAnnotations => 3,
        }
    }
}

/// Vertical anchor of positioned stamping text. Passed to
/// [`crate::EditableDoc::place_text_anchored`] and
/// [`crate::EditableDoc::place_paragraph`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum VerticalAnchor {
    /// `y` is the text baseline (historical [`crate::EditableDoc::place_text`]
    /// behavior). For paragraphs: the first line's baseline.
    #[default]
    Baseline,
    /// The text hangs from `y` (baseline at `y − ascent × size`, legacy layout engines
    /// `fixed-position layout` semantics). For paragraphs: top of the block.
    Top,
    /// The descender line rests on `y`. For paragraphs: bottom-pinned — the
    /// block grows upward from `y` by its real content height.
    Bottom,
    /// Top-anchored via the layout line box.
    LineTop,
    /// Bottom-anchored via the layout line box.
    LineBottom,
}

impl VerticalAnchor {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Baseline => 0,
            Self::Top => 1,
            Self::Bottom => 2,
            Self::LineTop => 3,
            Self::LineBottom => 4,
        }
    }
}

/// Vertical alignment of the text line inside a masked-text box. Passed to
/// [`crate::EditableDoc::masked_text_padded`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum VerticalAlign {
    /// The line hangs from the top edge (baseline at
    /// `y + height − ascent × size`, top line-alignment in rectangle-based text APIs).
    Top,
    /// Cap-height centering inside the box (historical
    /// [`crate::EditableDoc::masked_text`] behavior).
    #[default]
    Middle,
    /// The descender line rests on the bottom edge.
    Bottom,
}

impl VerticalAlign {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Top => 0,
            Self::Middle => 1,
            Self::Bottom => 2,
        }
    }
}

/// Coordinate space of the positioned stamping primitives. Passed to
/// [`crate::EditableDoc::set_stamp_space`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum StampSpace {
    /// The page's **displayed** space (historical default): coordinates
    /// compensate `/Rotate` so a `rotation_deg = 0` stamp reads upright.
    #[default]
    Visible,
    /// Raw PDF user space (legacy fixed-position layout/rotation
    /// semantics): no composition with the page's `/Rotate` or crop offset.
    Media,
}

impl StampSpace {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Visible => 0,
            Self::Media => 1,
        }
    }
}

/// How a rotated stamped image is anchored. Passed to
/// [`crate::EditableDoc::draw_image_anchored`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub enum ImageAnchor {
    /// Rotate the image about its lower-left corner at `(x, y)` (historical
    /// [`crate::EditableDoc::draw_image`] behavior).
    #[default]
    Corner,
    /// Land the rotated image's axis-aligned bounding box's lower-left corner
    /// at `(x, y)`.
    BoundingBox,
}

impl ImageAnchor {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::Corner => 0,
            Self::BoundingBox => 1,
        }
    }
}

/// PDF header version. Passed to [`crate::Document::set_version`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PdfVersion {
    V1_4,
    V1_5,
    V1_7,
    /// PDF 2.0 (ISO 32000-2).
    V2_0,
}

impl PdfVersion {
    pub(crate) fn code(self) -> i32 {
        match self {
            Self::V1_4 => 0,
            Self::V1_5 => 1,
            Self::V1_7 => 2,
            Self::V2_0 => 3,
        }
    }
}
